src/deploy/functions/build.ts

```diff
@@ -0,0 +1,397 @@
+import * as backend from "./backend";
+import * as proto from "../../gcp/proto";
+import { FirebaseError } from "../../error";
+
+/* The union of a customer-controlled deployment and potentially deploy-time defined parameters */
+export interface Build {
+  requiredAPIs: RequiredApi[];
+  endpoints: Record<string, Endpoint>;
+  params: Param[];
+}
+
+interface RequiredApi {
+  // The API that should be enabled. For Google APIs, this should be a googleapis.com subdomain
+  // (e.g. vision.googleapis.com)
+  api: string;
+
+  // A reason why this codebase requires this API.
+  // Will be considered required for all Extensions codebases. Considered optional for Functions
+  // codebases.
+  reason?: string;
+}
+
+// Defining a StringParam with { param: "FOO" } declares that "{{ params.FOO }}" is a valid
+// Expression<string> elsewhere.
+// Expression<number> is always an int. Float Params, list params, and secret params cannot be used in
+// expressions.
+// `Expression<Foo> == Expression<Foo>` is an Expression<boolean>
+// `Expression<boolean> ? Expression<T> : Expression<T>` is an Expression<T>
+type Expression<T extends string | number | boolean> = string;
+
+function resolveInt(from: number | Expression<number> | null): number {
+  if (from == null) {
+    return 0;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+function resolveString(from: string | Expression<string> | null): string {
+  if (from == null) {
+    return "";
+  }
+  return from;
+}
+
+function resolveBoolean(from: boolean | Expression<boolean> | null): boolean {
+  if (from == null) {
+    return false;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+// A service account must either:
+// 1. Be a project-relative email that ends with "@" (e.g. database-users@)
+// 2. Be a well-known shorthand (e..g "public" and "private")
+type ServiceAccount = string;
+
+// Trigger definition for arbitrary HTTPS endpoints
+interface HttpsTrigger {
+  // Which service account should be able to trigger this function. No value means "make public
+  // on create and don't do anything on update." For more, see go/cf3-http-access-control
+  invoker?: ServiceAccount | null;
+}
+
+// Trigger definitions for RPCs servers using the HTTP protocol defined at
+// https://firebase.google.com/docs/functions/callable-reference
+// eslint-disable-next-line
+interface CallableTrigger { }
+
+// Trigger definitions for endpoints that should be called as a delegate for other operations.
+// For example, before user login.
+interface BlockingTrigger {
+  eventType: string;
+}
+
+// Trigger definitions for endpoints that listen to CloudEvents emitted by other systems (or legacy
+// Google events for GCF gen 1)
+interface EventTrigger {
```

Comment:
Holy crap. After days of @taeold nerd snping me, I finally figured out how to use type programming to make a build just a type transformation on an Endpoint.

```
export type RecursiveKeyOf<T extends object> = {
  [Key in keyof T & (string | number)]:
    T[Key] extends any[]
    ? `${Key}`
    : T[Key] extends object
      ? `${Key}` | `${Key}.${RecursiveKeyOf<T[Key]>}`
      : `${Key}`;
}[keyof T & (string | number)];

export type HeadOf<T extends string> = [
  T extends `${infer Head}.${infer Tail}` ? Head : T
][number]

export type TailsOf<T extends string, Head extends string> = [
  T extends `${Head}.${infer Tail}` ? Tail : never
][number];

export type Parameterized<T> =
  T extends Array<infer A>
  ? Array<Parameterized<A>>
  : T extends object
    ? { [Key in keyof T & string]: Parameterized<T[Key]> }
    : T | string | null | undefined;

xport type Parameterize<T extends object, Paths extends RecursiveKeyOf<T>> = ParameterizeUnsafe<T, Paths>;

type ParameterizeUnsafe<T extends object, Paths extends string> = {
  [Key in keyof T]:
  Key extends Paths
  ? Parameterized<T[Key]>
  : Key extends HeadOf<Paths>
    ? ParameterizeUnsafe<T[Key] & object, TailsOf<Paths, Key>>
    : T[Key]
};

// Demo
interface Endpoint {
  platform: "gcfv1" | "gcfv2";
  entrypoint: number;
  eventTrigger: {
    eventType: number;
    filters: Record<string, string>;
  };
};
```

`Parameterized<Endpoint, "entrypoint" | "eventTrigger.eventType">` is equivalent to 

```
{
  platform: "gcfv1" | "gcfv2
  entrypoint: number | string | undefined | null
  eventTrigger: {
    eventType: number | string | undefined | null
    filters: Record<string, string>
  }
}
```